lib/kino/ecto.ex

```diff
@@ -0,0 +1,152 @@
+
+  @typedoc false
+  @type state :: %{
+          parent_monitor_ref: reference(),
+          repo: Ecto.Repo.t(),
+          queryable: Ecto.Queryable.t()
+        }
+
+  @doc """
+  Starts a widget process with the given queryable as
+  the data source.
+  """
+  @spec new(Ecto.Queryable.t(), Ecto.Repo.t()) :: t()
+  def new(queryable, repo) when is_atom(repo) do
+    unless queryable?(queryable) do
+      raise ArgumentError,
+            "expected a term implementing the Ecto.Queryable protocol, got: #{inspect(queryable)}"
+    end
+
+    parent = self()
+    opts = [repo: repo, queryable: queryable, parent: parent]
+
+    {:ok, pid} = DynamicSupervisor.start_child(Kino.WidgetSupervisor, {__MODULE__, opts})
+
+    %__MODULE__{pid: pid}
+  end
+
+  defp queryable?(term) do
+    Ecto.Queryable.impl_for(term) != nil
+  end
+
+  @doc false
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @impl true
+  def init(opts) do
+    repo = Keyword.fetch!(opts, :repo)
+    queryable = Keyword.fetch!(opts, :queryable)
+    parent = Keyword.fetch!(opts, :parent)
+
+    parent_monitor_ref = Process.monitor(parent)
+
+    {:ok, %{parent_monitor_ref: parent_monitor_ref, repo: repo, queryable: queryable}}
+  end
+
+  @impl true
+  def handle_info({:connect, pid}, state) do
+    {name, columns} =
+      if Table.ecto_schema?(state.queryable) do
+        name = state.queryable.__schema__(:source) |> to_string()
+        columns = state.queryable.__schema__(:fields) |> Table.keys_to_columns()
+        {name, columns}
+      else
+        {"Query", []}
+      end
+
+    features =
+      Kino.Utils.truthy_keys(
+        refetch: true,
+        pagination: true,
+        # If the user specifies custom select, the record keys
+        # are not valid "order by" fields, so we disable sorting
+        sorting: default_select_query?(state.queryable)
+      )
+
+    send(
+      pid,
+      {:connect_reply, %{name: name, columns: columns, features: features}}
+    )
+
+    {:noreply, state}
+  end
+
+  def handle_info({:get_rows, pid, rows_spec}, state) do
+    {total_rows, records} = get_records(state.repo, state.queryable, rows_spec)
+
+    {columns, keys} =
+      if Table.ecto_schema?(state.queryable) do
+        keys = state.queryable.__schema__(:fields)
+        {:initial, keys}
+      else
+        columns = Table.columns_for_records(records)
+        keys = Enum.map(columns, & &1.key)
+        {columns, keys}
+      end
+
+    rows = Enum.map(records, &Table.record_to_row(&1, keys))
+
+    send(pid, {:rows, %{rows: rows, total_rows: total_rows, columns: columns}})
+
+    {:noreply, state}
+  end
+
+  def handle_info({:DOWN, ref, :process, _object, _reason}, %{parent_monitor_ref: ref} = state) do
+    {:stop, :shutdown, state}
+  end
+
+  defp get_records(repo, queryable, rows_spec) do
+    count = repo.aggregate(queryable, :count)
+
+    query = from(q in queryable, limit: ^rows_spec.limit, offset: ^rows_spec.offset)
```

Comment:
Should we check the query has no limit and offset upfront? Otherwise this will raise. Alternatively we can call Query.exclude and remove both values upfront.

Reply:
Does it raise thought? It looks like the values are just overridden:

![image](https://user-images.githubusercontent.com/17034772/126660233-a523db20-a7ea-4750-a458-db1a86cc3fc5.png)

Reply:
Ah, I thought we raised. Good then!